File: AulaTPII/F217/GoF/Estruturais/Façade/facede01.js
// SubSistema 1
class SistemaPagamento{
    processarPagamento(valor){
        console.log(`Processamento pagamento no valor de R$: ${valor}`);
    }

}

// SubSistema 2
class SistemaLogistica{
    enviarProduto(destinatario){
        console.log(`Emviar produto para: ${destinatario}`);
    }

}

// SubSistema 3
class SistemaNotificacao{
    enviarEmail(destinatario, mensagem){
        console.log(`Emviar email para: ${destinatario}: ${mensagem}`); 
    }

}

// Fachada
class LojaOnline{
    constructor(){
        this.pagamento = new SistemaPagamento();
        this.logistica = new SistemaLogistica();
        this.notificacao = new SistemaNotificacao();
    }

    realizarCompra(produto, destinatario) {
        this.pagamento.processarPagamento(produto.preco);
        this.logistica.enviarProduto(destinatario);
        this.notificacao.enviarEmail(destinatario, `Seu pedido de ${produto.nome} foi enviado`);
    }
}



// Cliente - EXEMPLO DE USO ####################
const loja = new LojaOnline();

const produto = {
    nome: 'camiseta',
    preco: 49.99
}
const destinatario = 'alvaro@exemplo.com';

loja.realizarCompra(produto, destinatario);

File: AulaTPII/F217/GoF/Estruturais/Façade/facede01.py
# SubSistema 1
class SistemaPagamento:
    def processarPagamento(self, valor):
        print(f"processaento pagemento no valor de R$ {valor}")

# SubSistema 2
class SistemaLogistica:
    def enviarProduto(self, destinatario):
        print(f"Enviar produto para: {destinatario}")
# SubSistema 3
class SistemaNotificacao:
    def enviarEmail(self, destinatario, mensagem):
        print(f"Enviar email para: {destinatario}: {mensagem}")

# Fachada
class LojaOnline:
    def __init__(self):
        self.pagamento = SistemaPagamento()
        self.logistica = SistemaLogistica()
        self.notificacao = SistemaNotificacao()
    
    def realizarPagamento(self, produto, destinatario):
        self.pagamento.processarPagamento(produto['preco'])
        self.logistica.enviarProduto(destinatario)
        self.notificacao.enviarEmail(destinatario, f"Seu pedido de {produto['nome']} foi enviado")

loja = LojaOnline()

produto = {
    'nome': 'camiseta',
    'preco': 49.99
}
destinatario = 'alvaro@exemplo.com'

loja.realizarPagamento(produto, destinatario)